system-helper/flatpak-system-helper.c

```diff
@@ -141,6 +301,177 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
+      g_autoptr(FlatpakSystemHelperOngoingPull) pull_tmp = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      G_LOCK (ongoing_pulls);
+      for (guint i = 0; i < ongoing_pulls->len; i++)
+        {
+          pull_tmp = g_ptr_array_index (ongoing_pulls, i);
+          if (pull_tmp->uid == uid)
+            {
+              uid_pull_count++;
+              // TODO: If there is a already a partial pull available, return that.
+            }
+        }
+      G_UNLOCK (ongoing_pulls);
+      pull_tmp = NULL;
+
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      root_dir = g_mkdtemp_full (g_build_filename (flatpak_dir, "repo", "tmp", "flatpak-cache-XXXXXX", NULL), 0755);
+      user_dir = g_strdup_printf ("%s-bindfs", root_dir);
+      if (g_mkdir_with_parents (user_dir, 0755) == -1)
+        {
+          glnx_throw_errno_prefix (&error, "Can't create directory for bindfs mount at %s", user_dir);
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      pull_tmp = system_helper_ongoing_pull_new (object, invocation, uid, user_dir);
+
+      arg_user = g_strdup_printf ("--force-user=%"G_GUINT32_FORMAT, uid);
+      pull_tmp->bindfs = g_subprocess_new (G_SUBPROCESS_FLAGS_NONE,
+                                           &error,
+                                           bindfs_path,
+                                           arg_user,"-f",
+                                           "--create-for-user=root",
+                                           "--create-for-group=root",
+                                           root_dir, user_dir, NULL);
+      if (pull_tmp->bindfs == NULL)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_SPAWN_FAILED,
+                                                 "Error spawning bindfs for user:%"G_GUINT32_FORMAT" at: %s", uid, user_dir);
+          return TRUE;
+        }
+
+      pull_tmp->root_dir_name = g_steal_pointer (&root_dir);
+
+      new_pull = g_steal_pointer (&pull_tmp);
+      G_LOCK (ongoing_pulls);
+      g_ptr_array_add (ongoing_pulls, new_pull);
+      G_UNLOCK (ongoing_pulls);
+
+      g_subprocess_wait_check_async (new_pull->bindfs, NULL, subprocess_wait_check_cb, new_pull);
```

Comment:
To solve the problem where the callback may be called after `new_pull`:

* Add a `GCancellable` to the OngoingPull struct
* Cancel it at the very start of system_helper_ongoing_pull_free
* Pass it here
* In `subprocess_wait_check_cb`, check if `g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED)` and if so take no action.